### PR TITLE
3.0.0

### LINKED ISSUE #253: add CTF grouping to 3d init protocol
Looks missing..

### PR DESCRIPTION
**Pull request by azazellochg**
- Update extra labels list for particles: 
```
PARTICLE_EXTRA_LABELS = [
    'rlnCtfBfactor',
    'rlnCtfScalefactor',
    'rlnPhaseShift',
    'rlnGroupName',
    'rlnGroupNumber',
    'rlnRandomSubset'
]
```
- set PARTICLE_EXTRA_LABELS inside ClassesLoader and refine3d _updateParticle
- Fill random subset by default. Is there any use case when we do NOT want it?
- Autoscale input vol refs with relion not xmipp
- Add volume resize protocol with a test
- Fix #253, #256, #258, #269
- Add missing ProtRelionExportParticles to protocol tree
- Fix default values for missing ctf attrs
- Use abs paths when creating symlinks - important for making scratch work
- Add warning about dm4 gain flip by xmipp
- Fix error in multi-body protocol
- Optimize imports

### CODE REVIEW

**Comment by azazellochg** [comment]
I know it makes conversion slower, so would welcome any comments.

File: relion/convert/convert31.py [hunk 1]
```diff
@@ -657,3 +655,22 @@ def _get(label):
...
+            self._extraLabels = [l for l in extraLabels if row.hasColumn(l)]
+
+            for label in self._extraLabels:
+                setattr(particle, '_' + label,
+                        ObjectWrap(getattr(row, label)))
+
+            self._first = False
+
+        for label in self._extraLabels:
+            getattr(particle, '_%s' % label).set(getattr(row, label))
```
Comment (azazellochg) [inline relion/convert/convert31.py:676]: check how this affects performance

File: relion/convert/__init__.py [hunk 2]
```diff
@@ -144,27 +145,12 @@ def _updateParticle(self, item, row):
         item.setClassId(row.rlnClassNumber)
         self._reader.setParticleTransform(item, row)
 
-        # Try to create extra objects only once if item is reused
```
Comment (azazellochg) [inline relion/convert/__init__.py:147]: groupname/number is now in extraLabels which is always parsed if exists for the first part / row. I do not see the use of extra ML-related labels that decrease overall conversion performance and also create unnecessary Scipion objects (Float)

File: relion/protocols/protocol_base.py [hunk 3]
```diff
@@ -1131,7 +1122,7 @@ def _getIterClasses(self, it, clean=False):
         if clean:
             pwutils.cleanPath(data_classes)
 
-        if not os.path.exists(data_classes):
+        if not pwutils.exists(data_classes):
```
Comment (delarosatrevin) [inline relion/protocols/protocol_base.py:1125]: This usage is misleading...the pwutils does not have an 'exists' method, it is somewhere imported from os.path, but this is implementation details, so it should not be used. It is better to use os.path.exists directly. Please check all occurrences of this.

File: relion/protocols/protocol_base.py [hunk 4]
```diff
@@ -1148,8 +1139,8 @@ def _getIterData(self, it, **kwargs):
         """ Sort the it??.data.star file by the maximum likelihood. """
         data_sqlite = self._getFileName('data_scipion', iter=it)
 
-        if not os.path.exists(data_sqlite):
-            iterImgSet = pwem.objects.SetOfParticles(filename=data_sqlite)
+        if not pwutils.exists(data_sqlite):
```
Comment (delarosatrevin) [inline relion/protocols/protocol_base.py:1142]: Same here.

File: relion/protocols/protocol_bayesian_polishing.py [hunk 5]
```diff
@@ -338,7 +339,7 @@ def _params(label, *params):
         else:
             outputFn = None
             for fn in ['opt_params.txt', 'opt_params_all_groups.txt']:
-                if os.path.exists(self._getExtraPath(fn)):
+                if pwutils.exists(self._getExtraPath(fn)):
```
Comment (delarosatrevin) [inline relion/protocols/protocol_bayesian_polishing.py:342]: use os.path.exists

File: relion/protocols/protocol_compress_movies.py [hunk 6]
```diff
@@ -160,6 +160,6 @@ def getInputGain(self):
         inputGainProt = self.inputGainProt.get()
         if inputGainProt is not None:
             inputGain = inputGainProt.getOutputGain()
-            if os.path.exists(inputGain):
+            if pwutils.exists(inputGain):
```
Comment (delarosatrevin) [inline relion/protocols/protocol_compress_movies.py:163]: os.path.exists

File: relion/protocols/protocol_export_ctf.py [hunk 7]
```diff
@@ -87,7 +86,7 @@ def writeCtfStarStep(self):
         micSet = SetOfMicrographs(filename=':memory:')
 
         psd = inputCTF.getFirstItem().getPsdFile()
-        hasPsd = psd and os.path.exists(psd)
+        hasPsd = psd and pwutils.exists(psd)
```
Comment (delarosatrevin) [inline relion/protocols/protocol_export_ctf.py:89]: os.path.exists

File: relion/protocols/protocol_refine3d.py [hunk 8]
```diff
@@ -179,8 +180,11 @@ def _fillDataFromIter(self, imgSet, iteration):
     def _updateParticle(self, particle, row):
         self.reader.setParticleTransform(particle, row)
 
-        if not hasattr(particle, '_rlnRandomSubset'):
-            particle._rlnRandomSubset = Integer()
-
-        particle._rlnRandomSubset.set(row.rlnRandomSubset)
-
+        if self.reader._first:
```
Comment (delarosatrevin) [inline relion/protocols/protocol_refine3d.py:183]: This should be handled in the reader...seems like duplicated code...and make use of the internal _first property. I can help with this

File: relion/tests/test_protocols_relion3.py [hunk 9]
```diff
@@ -27,10 +27,10 @@
 import os
 
 import pyworkflow.tests as pwtests
+from pyworkflow.utils import copyTree, join, magentaStr
```
Comment (delarosatrevin) [inline relion/tests/test_protocols_relion3.py:30]: join is not from pwutils, it comes from os.path. The same comment about os.path.exists applies here.

**Review by delarosatrevin — CHANGES_REQUESTED** [review]
Some changes requested...I will continue looking at the code and testing new protocols

File: relion/convert/convert31.py [hunk 10]
```diff
@@ -479,7 +487,7 @@ def __init__(self, **kwargs):
         """
         """
         ReaderBase.__init__(self, **kwargs)
-        self._first = False
+        self._first = True
```
Comment (delarosatrevin) [inline relion/convert/convert31.py:490]: I think we don't need these property

File: relion/protocols/protocol_base.py (hunk 3 above)
Comment (azazellochg) [reply relion/protocols/protocol_base.py:1125]: oh no, I've been replacing it for years everywhere...

**Comment by azazellochg** [comment]
Thanks a lot, @delarosatrevin , changes done!

**Pull request merged**